File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int helper(string& s, string& t, int ind1, int ind2,
               vector<vector<int>>& dp) {
        if (ind2 < 0)
            return 1;
        if (ind1 < 0)
            return 0;
        if (dp[ind1][ind2] != -1)
            return dp[ind1][ind2];

        if (s[ind1] == t[ind2]) {
            return dp[ind1][ind2] = helper(s, t, ind1 - 1, ind2 - 1, dp) +
                                    helper(s, t, ind1 - 1, ind2, dp);
        }
        return dp[ind1][ind2] = helper(s, t, ind1 - 1, ind2, dp);
    }

    int numDistinct(string s, string t) {
        int ind1 = s.length() - 1;
        int ind2 = t.length() - 1;
        vector<vector<int>> dp(ind1 + 1, vector<int>(ind2 + 1, -1));

        return helper(s, t, ind1, ind2, dp);
    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.java
class Solution {
 int helper(String s, String t, int index1, int index2, Integer[][] dp) {
        // base...
        if (index2 < 0)
            return 1;
        if (index1 < 0)
            return 0;
        if (dp[index1][index2] != null)
            return dp[index1][index2];

        if (s.charAt(index1) == t.charAt(index2)) {
            return dp[index1][index2] = helper(s, t, index1 - 1, index2 - 1, dp) + helper(s, t, index1 - 1, index2, dp);
        } else {
            return dp[index1][index2] = helper(s, t, index1 - 1, index2, dp);
        }
    }

    public int numDistinct(String s, String t) {
        int index1 = s.length();
        int index2 = t.length();
        Integer[][] dp = new Integer[index1][index2];

        return helper(s, t, index1 - 1, index2 - 1, dp);
    }
}